File: scala-patterns/patterns/src/Exp.scala
import Elt._
import Type._

// object Exp {

//   case class Exp[a, r](r: PreExp[r])(implicit repr: EltRepr[a, r])

//   sealed trait PreExp[a]

//   case class Constant[a]  (s: ScalarType[a], x: a)      extends PreExp[a]
//   case class Pair    [a,b](x: PreExp[a], y: PreExp[b])  extends PreExp[(a,b)]
//   case class UExp         ()                            extends PreExp[Unit]
//   case class PrjL    [a,b](x: PreExp[(a,b)])            extends PreExp[a]
//   case class PrjR    [a,b](x: PreExp[(a,b)])            extends PreExp[b]

//   case class Undef   [a]  (x: TypeR[a])                 extends PreExp[a]
//   case class Match   [a]  (t: TagR[a], x: PreExp[a])    extends PreExp[a]
//   case class Case    [a,b](x: PreExp[a], js: List[(TagR[a], PreExp[b])]) extends PreExp[b]

//   case class Var     [a]  (i: Idx[a])                   extends PreExp[a]
//   case class Let     [a,b](i: Idx[a], x: PreExp[a], y: PreExp[b]) extends PreExp[b]

//   case class Idx[t](x: TypeR[t], t: String)

// }

object Exp {

  // sealed trait Exp[T]

  case class Exp[T,R](e: PreExp[R])(implicit val elt: Elt[T,R])

  def plus[R](x: Exp[Int,Int], y: Exp[Int,Int]) = Exp(Plus(x.e, y.e))

  sealed trait PreExp[+R]

  // case class Rep[T,R] (e: Exp[T,R])(implicit elt: Elt[T,R]) extends PreExp[R]
  // case class Surface[T,R] (e: PreExp[R])(implicit elt: Elt[T,R]) extends PreExp[T]

  case class Const   [T,R] (e: T)        (implicit elt: Elt[T,R]) extends PreExp[R]
  // case class Tuple   [T,R] (e: PreExp[R])(implicit elt: Elt[T,R]) extends PreExp[T]

  case class Pair    [A,B](x: PreExp[A], y: PreExp[B])  extends PreExp[(A,B)]
  case class UExp         ()                            extends PreExp[Unit]
  case class PrjL    [A,B](x: PreExp[(A,B)])            extends PreExp[A]
  case class PrjR    [A,B](x: PreExp[(A,B)])            extends PreExp[B]

  case class Undef   [T,R] ()(implicit elt: Elt[T,R])     extends PreExp[R]
  case class Match   [R] (t: TagR[R], x: PreExp[R]) extends PreExp[R]
  case class Case    [a,b](x: PreExp[a], js: List[(TagR[a], PreExp[b])]) extends PreExp[b]

  // case class Var     [a]  (i: Idx[a])                   extends PreExp[a]
  // case class Let     [a,b](i: Idx[a], x: PreExp[a], y: PreExp[b]) extends PreExp[b]

  // case class Idx[t](x: TypeR[t], t: String)

  case class Unroll[T,R](x: Exp[Rec[T],Rec[R]]) extends PreExp[R]
  case class Roll  [T,R](x: Exp[T,R])           extends PreExp[Rec[R]]

  case class Plus(x: PreExp[Int], y: PreExp[Int]) extends PreExp[Int]

}

File: haskell-surface/src/Exp.hs
{-# LANGUAGE AllowAmbiguousTypes #-}
{-# LANGUAGE GADTs               #-}
{-# LANGUAGE LambdaCase          #-}
{-# LANGUAGE ScopedTypeVariables #-}

module Exp where

import Elt
import Rec
import Trace
import Tuple
import Type
import Debug

import Data.Dynamic
import Data.Map                                           ( Map )
import Data.Text                                          ( Text )
import qualified Data.Map                                 as Map
import qualified Data.Text                                as T


data Exp a where
  -- STLC with explicit let-binding. We don't bother with the type-level environment.
  -- Values are stored in representation format
  Const     :: Elt a => EltR a -> Exp a
  Var       :: Idx t -> Exp t
  Let       :: Idx a -> Exp a -> Exp b -> Exp b -- letrec
  App       :: Exp (a -> b) -> Exp a -> Exp b
  Lam       :: Idx a -> Exp b -> Exp (a -> b)

  -- Add generic product types
  Tuple     :: IsTuple t => Tuple (TupleR t) -> Exp t
  Prj       :: IsTuple t => TupleIdx (TupleR t) e -> Exp t -> Exp e

  -- Recursive types
  Unroll    :: Exp (Rec a) -> Exp a
  Roll      :: Exp a -> Exp (Rec a)

  -- Pattern matching
  Undef     :: TypeR (EltR a) -> Exp a
  Match     :: TraceR (EltR a) -> Exp a -> Exp a
  Case      :: Elt a => Exp a -> [(TraceR (EltR a), Exp b)] -> Exp b

  -- PrimOps
  Add       :: Exp Int -> Exp Int -> Exp Int
  Eq        :: Exp Int -> Exp Int -> Exp Bool

-- Tuples are heterogeneous lists using () and (,)
--
data Tuple t where
  Unit :: Tuple ()
  Exp  :: Exp a -> Tuple a
  Pair :: Tuple a -> Tuple b -> Tuple (a, b)

data TupleIdx s t where
  PrjZ ::                 TupleIdx t      t
  PrjL :: TupleIdx l t -> TupleIdx (l, r) t
  PrjR :: TupleIdx r t -> TupleIdx (l, r) t


-- Very unsafe variable bindings!
--
data Idx t where
  Idx :: Typeable t => Text -> Idx t

type Env = Map Text Dynamic


-- Standard type-safe evaluator
--
eval :: Exp a -> a
eval = evalExp Map.empty

evalExp :: Env -> Exp a -> a
evalExp env = \case
  Const c         -> toElt c
  Var ix          -> lookupEnv ix env
  Let (Idx v) a b -> let env' = Map.insert v (toDyn (evalExp env' a)) env
                      in evalExp env' b
  Lam (Idx v) b   -> \a -> evalExp (Map.insert v (toDyn a) env) b
  App f x         -> (evalExp env f) (evalExp env x)
  Tuple t         -> toTup $ evalTup env t
  Prj tix t       -> evalPrj tix (fromTup (evalExp env t))
  Unroll e        -> let Rec x = evalExp env e in x
  Roll e          -> Rec (evalExp env e)
  Undef{}         -> error "evalExp: undef"
  Match _ e       -> if dEBUG
                       then error "evalExp: match"
                       else evalExp env e
  Case x xs       -> evalExp env $ lookupCase (fromElt (evalExp env x)) xs
  --
  Add x y         -> evalExp env x + evalExp env y
  Eq x y          -> evalExp env x == evalExp env y

evalTup :: Env -> Tuple t -> t
evalTup env = \case
  Unit     -> ()
  Pair a b -> (evalTup env a, evalTup env b)
  Exp e    -> evalExp env e

evalPrj :: TupleIdx t e -> t -> e
evalPrj PrjZ e = e
evalPrj (PrjL t) (l, _) = evalPrj t l
evalPrj (PrjR t) (_, r) = evalPrj t r

lookupCase :: a -> [(TraceR a, Exp b)] -> Exp b
lookupCase val = go
  where
    go [] = error "lookupCase: unmatched case"
    go ((tag,cont):rest)
      | eqTrace tag val = cont
      | otherwise       = go rest

lookupEnv :: forall t. Idx t -> Env -> t
lookupEnv (Idx nm) env
  | Just v  <- Map.lookup nm env
  , Just v' <- fromDynamic v
  = v'
  | otherwise
  = error ("lookupEnv: not found: " ++ T.unpack nm)

eqTrace :: TraceR a -> a -> Bool
eqTrace TraceRunit         ()      = True
eqTrace (TraceRrec _)      (Rec _) = True
eqTrace (TraceRprim _)     _       = True
eqTrace (TraceRundef _)    _       = True
eqTrace (TraceRtag tag ta) (t, a)  = t == tag && eqTrace ta a
eqTrace (TraceRpair ta tb) (a, b)  = eqTrace ta a && eqTrace tb b

